### PR TITLE
feat(normalization): Trim logger smartly

### LINKED ISSUE #1827: Increase character limit for the logger
A customer gets it's value from the logger but it's been capped at 64 which truncates the information.
Is there a reason for this or can we up the limit to a fair amount like for instance the TagValue => 200

`"_meta": {
    "logger": {
      "": {
        "rem": [
          [
            "!limit",
            "s",
            61,
            64
          ]
        ],
        "len": 86
      }
    }
  },`

can be found in this [file ](https://github.com/getsentry/relay/blob/6ebaf355ef20a2c10c5d1ced99f5a296b31c396c/relay-general/src/processor/attrs.rs#L117-L138)

### PR DESCRIPTION
**Pull request by iker-barriocanal**
This PR changes how we trim loggers for events coming from Java platforms. Chars from the end are now prioritized (so the string is trimmed from the beginning), and it's also attempted to trim full words in `.`s.

We are only interested in doing this for Java at this moment; because loggers in Java are fully qualified names (including the package, so potentially very long strings) where the class name at the end is more important, and because it's the only platform we've received this type of feedback on.

Resolves https://github.com/getsentry/relay/issues/1827.

### Examples

These examples are from the tests, see the diff for more examples and/or edge cases.

```txt
# Before
this_is-way_too_long-and_we_only_have_one_word-so_we_cant_smart_trim

# After
*is-way_too_long-and_we_only_have_one_word-so_we_cant_smart_trim
```

```txt
# Before
super_out.this_is_already_out_too.this_part-is-kept.this_right_here-is_a-very_long_word

# After
*.this_part-is-kept.this_right_here-is_a-very_long_word
```

### CODE REVIEW

File: relay-general/src/store/normalize.rs [hunk 1]
```diff
@@ -1001,6 +1002,78 @@ impl<'a> Processor for NormalizeProcessor<'a> {
...
         Ok(())
     }
+
+    fn process_event_logger(
+        &mut self,
+        event_logger: &mut EventLogger,
+        meta: &mut Meta,
+        state: &ProcessingState,
+    ) -> ProcessingResult {
+        event_logger.process_child_values(self, state)?;
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]: Why even call `process_child_values`? `EventLogger` will always be a leaf node AFAIK.

File: relay-general/src/store/normalize.rs [hunk 2]
```diff
@@ -1001,6 +1002,78 @@ impl<'a> Processor for NormalizeProcessor<'a> {
...
+            event_logger.set_logger(trimmed_logger.to_owned());
+        }
+
+        Ok(())
+    }
+}
+
+static LOGGER_DELIMITER_PATTERN: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"\.").expect("Invalid logger separator Regex"));
+static LOGGER_TRIMMED_PREFIX: &str = "*";
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]:
```suggestion
const LOGGER_TRIMMED_PREFIX: &str = "*";
```

File: relay-general/src/store/normalize.rs [hunk 3]
```diff
@@ -1001,6 +1002,78 @@ impl<'a> Processor for NormalizeProcessor<'a> {
...
+
+fn shorten_logger(logger: &str) -> String {
+    let mut tokens = LOGGER_DELIMITER_PATTERN.split(logger).collect_vec();
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]: Using regex here seems like overkill to me.
```suggestion
    let mut tokens = logger.split(LOGGER_DELIMITER_CHAR).collect_vec();
```

File: relay-general/src/store/normalize.rs [hunk 4]
```diff
@@ -1001,6 +1002,78 @@ impl<'a> Processor for NormalizeProcessor<'a> {
...
+    tokens.reverse(); // Prioritize keeping chars from the end
+
+    let logger_len = bytecount::num_chars(logger.as_bytes());
+    let target_size = logger_target_size(tokens);
+    let target_idx = logger_len - target_size;
+
+    LOGGER_TRIMMED_PREFIX.to_owned() + &logger[target_idx..]
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]: `logger[target_idx..]` could panic if `target_idx` is not a character boundary.

E.g. `&"Löwe"[2..]` will result in 

```
thread 'main' panicked at 'byte index 2 is not a char boundary; it is inside 'ö' (bytes 1..3) of `Löwe`'
```

File: relay-general/src/store/normalize.rs [hunk 5]
```diff
@@ -1001,6 +1002,78 @@ impl<'a> Processor for NormalizeProcessor<'a> {
...
+    if MaxChars::Logger.limit() < logger_prefix_len {
+        return 0;
+    }
+
+    let max = MaxChars::Logger.limit() - logger_prefix_len;
+    let mut cummulative = 0;
+
+    for token in tokens {
+        // Consider one slot for the delimiter between tokens.
+        let current = bytecount::num_chars(token.as_bytes()) + 1;
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]: It's a good idea to use `num_chars` for limiting the string, but the computed `target_size` that is used for slicing the string should be in bytes. Best add some unit tests with strings that contain a lot of special characters to see the effects.

**Review by jjbayer — COMMENTED** [review]
I like the idea of a dedicated `EventLogger` type, _but_ if we want to make this normalization dependent on other fields like `platform`, I think it's easier to add a function to `light_normalize_event` instead, similar to `normalize_logentry`.

**Comment by iker-barriocanal** [comment]
> but if we want to make this normalization dependent on other fields like `platform`, I think it's easier to add a function to `light_normalize_event` instead, similar to `normalize_logentry`

@jjbayer if we need to do this platform-dependent, I was thinking of doing it in the normalize processor instead. I don't think we need it as part of the light normalization.

File: relay-general/src/store/normalize.rs [hunk 6]
```diff
@@ -806,6 +806,19 @@ impl<'a> Processor for NormalizeProcessor<'a> {
             event.client_sdk.set_value(self.get_sdk_info());
         }
 
+        if event.platform.as_str() == Some("java") {
+            if let Some(event_logger) = event.logger.value_mut() {
+                let trimmed_logger = event_logger.as_str().trim();
+                let logger_len = bytecount::num_chars(trimmed_logger.as_bytes());
+
+                if logger_len > MaxChars::Logger.limit() {
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]: You do the same check in `shorten_logger`, so I would remove it here.

File: relay-general/src/store/normalize.rs [hunk 7]
```diff
@@ -806,6 +806,19 @@ impl<'a> Processor for NormalizeProcessor<'a> {
             event.client_sdk.set_value(self.get_sdk_info());
         }
 
+        if event.platform.as_str() == Some("java") {
+            if let Some(event_logger) = event.logger.value_mut() {
+                let trimmed_logger = event_logger.as_str().trim();
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]: Can we move this basic trimming inside `shorten_logger`?

File: relay-general/src/store/normalize.rs [hunk 8]
```diff
@@ -1005,6 +1018,68 @@ impl<'a> Processor for NormalizeProcessor<'a> {
...
+/// String.
+///
+/// To shorten the logger, all extra chars that don't fit into the maximum limit
+/// are removed, from the beginning of the logger.  Then, if the remaining
+/// substring contains a `.` somewhere but in the end, all chars until `.`
+/// (exclusive) are removed.
+///
+/// Additionally, the new logger is prefixed with `*`, to indicate it was
+/// shortened.
+fn shorten_logger(logger: &str) -> String {
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]: nit: I would rename this to `trim_logger`, since that is the word we use in other places (e.g. `TrimmingProcessor`).

File: relay-general/src/store/normalize.rs [hunk 9]
```diff
@@ -1005,6 +1018,68 @@ impl<'a> Processor for NormalizeProcessor<'a> {
...
+    let logger_len = bytecount::num_chars(logger.as_bytes());
+    if logger_len <= MaxChars::Logger.limit() {
+        return logger.to_owned();
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]: You could change the interface of this function to consume a `String`, then you could return `logger` here without extra allocation:

`fn shorten_logger(logger: String) -> String {`

File: relay-general/src/store/normalize.rs [hunk 10]
```diff
@@ -1005,6 +1018,68 @@ impl<'a> Processor for NormalizeProcessor<'a> {
...
+    }
+
+    let mut tokens = logger.split("").collect_vec();
+    // Remove empty str tokens from the beginning and end.
+    tokens.pop();
+    tokens.reverse(); // Prioritize chars from the end of the string.
+    tokens.pop();
+
+    remove_logger_extra_chars(&mut tokens);
+    remove_logger_word(&mut tokens);
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]: Shouldn't this be called only if `remove_logger_extra_chars` cut a word in half? If my logger is `foo.bar`, wouldn't `remove_logger_word` crop it to `.bar`?

File: relay-general/src/store/normalize.rs [hunk 11]
```diff
@@ -1005,6 +1018,68 @@ impl<'a> Processor for NormalizeProcessor<'a> {
...
+
+    tokens.reverse();
+    "*".to_owned() + &tokens.iter().join("")
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]:
```suggestion
    tokens.join("*")
```

Comment (iker-barriocanal) [reply relay-general/src/store/normalize.rs]: `join` is the contrary operation to `split`. We only want the `*` at the beginning to indicate the logger was trimmed, not between each token (where a token is a character).

File: relay-general/src/store/normalize.rs (hunk 9 above)
Comment (iker-barriocanal) [reply relay-general/src/store/normalize.rs]: It's a good suggestion. However, I've moved the trimming to this function too and `trim` returns an `str`, so we still have an extra allocation in this case.

File: relay-general/src/store/normalize.rs (hunk 10 above)
Comment (iker-barriocanal) [reply relay-general/src/store/normalize.rs]: That's an interesting case I didn't think about. Should be addressed in db4d016.

File: relay-general/src/store/normalize.rs (hunk 8 above)
Comment (iker-barriocanal) [reply relay-general/src/store/normalize.rs]: Unless there're strong opinions about changing that name, I'd rather not. The reason is that what this method does isn't really trimming:

> In computer programming, trimming or stripping is a string manipulation in which leading and trailing whitespace is removed from a string.

We don't trim here. We remove a non-whitespace substring from the string and insert a `*`. I'd like to avoid any future confusion coming from the wording.

File: relay-general/src/store/normalize.rs (hunk 11 above)
Comment (jjbayer) [reply relay-general/src/store/normalize.rs]: Whoops, I guess I meant something like
```suggestion
    format!("*{}", tokens.join(""))
```

File: relay-general/src/store/normalize.rs [hunk 12]
```diff
@@ -808,13 +808,9 @@ impl<'a> Processor for NormalizeProcessor<'a> {
...
         if event.platform.as_str() == Some("java") {
             if let Some(event_logger) = event.logger.value_mut() {
-                let trimmed_logger = event_logger.as_str().trim();
-                let logger_len = bytecount::num_chars(trimmed_logger.as_bytes());
-
-                if logger_len > MaxChars::Logger.limit() {
-                    *event_logger = shorten_logger(trimmed_logger);
-                } else if trimmed_logger != event_logger.as_str() {
-                    *event_logger = trimmed_logger.to_owned();
+                let shortened = shorten_logger(event_logger.clone());
```
Comment (jjbayer) [inline relay-general/src/store/normalize.rs]: nit: It must be possible to prevent this clone. Maybe change the signature to `shorten_logger(&mut String) -> ()`? Then the check below becomes unnecessary.

**Review by jjbayer — APPROVED** [review]

File: relay-general/src/store/normalize.rs (hunk 12 above)
Comment (iker-barriocanal) [reply relay-general/src/store/normalize.rs]: If we change the signature to accept a `&mut String`, we remove a cloning here. However, in cases where the logger is smaller than the max size, we will always perform an allocation (if we clone the String, not always) which is not the case for cloning; plus, making the code less readable (why do we need to pass a mutable string?). So optimizing to remove extra allocations doesn't always happen.

I don't have data to back up the following, but my guess is that the performance improvement is not noticeable. This makes me optimize how the code is used instead.

Of course, I may be missing something that makes this possible _and_ readable. If that's the case, I'm happy to update the code. If not, I'll go ahead and merge it like this.

Comment (jjbayer) [reply relay-general/src/store/normalize.rs]: > I may be missing something that makes this possible and readable.

Opened a little draft PR, let me know what you think: https://github.com/getsentry/relay/pull/1950

**Comment by jjbayer** [comment]
@iker-barriocanal One thing I missed during review: Should we add a `Meta` remark when we change the string? Something like

https://github.com/getsentry/relay/blob/1c47b22e2ddee545a9b76e09f8eca450e83b8965/relay-general/src/store/trimming.rs#L424-L429

**Review by jjbayer — APPROVED** [review]
